File: group_anagrams/group_anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
  const anagramMap = new Map<string, string[]>();
  for (const str of strs) {
    const sortedKey = str.split("").sort().join("");
    if (!anagramMap.has(sortedKey)) {
      anagramMap.set(sortedKey, []);
    }
    anagramMap.get(sortedKey)!.push(str);
  }
  return Array.from(anagramMap.values());
}

function groupAnagramsOptimized(strs: string[]): string[][] {
  const anagramMap = new Map<string, string[]>();
  for (const str of strs) {
    const count = new Array(26).fill(0);
    for (const char of str) {
      count[char.charCodeAt(0) - 'a'.charCodeAt(0)]++;
    }
    const key = count.join('#');
    if (!anagramMap.has(key)) {
      anagramMap.set(key, []);
    }
    anagramMap.get(key)!.push(str);
  }
  return Array.from(anagramMap.values());
}

const strs = ["eat", "tea", "tan", "ate", "nat", "bat"];

const result1 = groupAnagrams(strs);
console.log("Grouped Anagrams: ", result1);

const result2 = groupAnagramsOptimized(strs);
console.log("Grouped Anagrams (Optimized): ", result2);

File: group_anagrams/group_anagrams.py
from collections import defaultdict

def group_anagrams(strs):
    anagram_map = {}
    for str in strs:
        sorted_str = ''.join(sorted(str))
        if(sorted_str in anagram_map):
            anagram_map[sorted_str].append(str)
        else:
            anagram_map[sorted_str] = [str]
    return list(anagram_map.values())

def group_anagrams_using_defaultdict(strs):
    anagram_map = defaultdict(list)
    for str in strs:
        sorted_str = ''.join(sorted(str))
        anagram_map[sorted_str].append(str)
    return list(anagram_map.values())

strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
result = group_anagrams(strs)
print("Grouped anagrams:", result)  # Output: [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]